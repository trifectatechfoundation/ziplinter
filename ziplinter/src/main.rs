use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
