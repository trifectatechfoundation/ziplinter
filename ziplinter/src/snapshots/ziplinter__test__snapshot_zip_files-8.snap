---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "directory_headers": [
    {
      "comment": "",
      "compressed_size": 0,
      "crc32": 0,
      "creator_version": {
        "host_system": "Unix",
        "version": 30
      },
      "disk_nbr_start": 0,
      "external_attrs": 2175008768,
      "extra": "UT\u0005\u0000\u0003ç╧\u0000Zux\u000b\u0000\u0001\u0004Φ\u0003\u0000\u0000\u0004Φ\u0003\u0000\u0000",
      "flags": 2048,
      "header_offset": 0,
      "internal_attrs": 0,
      "method": "Store",
      "modified": {
        "date": 19302,
        "time": 26926
      },
      "name": "Σ╕ûτòî",
      "reader_version": {
        "host_system": "MsDos",
        "version": 10
      },
      "uncompressed_size": 0
    }
  ],
  "encoding": "Utf8",
  "entries": [
    {
      "accessed": null,
      "comment": "",
      "compressed_size": 0,
      "crc32": 0,
      "created": null,
      "flags": 2048,
      "gid": 1000,
      "header_offset": 0,
      "method": "Store",
      "mode": 420,
      "modified": "2017-11-06T21:09:27Z",
      "name": "世界",
      "reader_version": {
        "host_system": "MsDos",
        "version": 10
      },
      "uid": 1000,
      "uncompressed_size": 0
    }
  ],
  "eocd": {
    "dir": {
      "inner": {
        "comment": [],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 64,
        "directory_records": 1,
        "directory_size": 76,
        "disk_nbr": 0
      },
      "offset": 140
    },
    "dir64": null,
    "global_offset": 0
  },
  "size": 162
}
